### PR TITLE
Asynchronous endpoint selection

### LINKED ISSUE #1910: HealthCheckedEndpointGroup should probably wait for connectTimeout before throwing exception.
Currently, `HttpHealthCheckedEndpointGroup` will throw an `EndpointGroupException` immediately when a request is attempted when there are no endpoints. It would be more intuitive and simpler to use if it was possible for the request to wait until `ClientFactory.connectTImeout()` before throwing the exception, allowing a health checked endpoint group to have similar connection behavior as just a single endpoint.

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

Currently, sending a request will fail immediately with an
`EmptyEndpointGroupException` when the target `EndpointGroup` is empty
without waiting at all.

If we could wait until the `EndpointGroup` has at least one `Endpoint`,
we will be able to avoid the `EmptyEndpointGroupException` at least for
the case where the `Endpoint`s disappear temporarily.

Modifications:

- Rename `EndpointSelector.select()` to `selectNow()`.
- Add `EndpointSelector.select()` that performs asynchronous endpoint
  selection with timeout.
- Add `AbstractEndpointSelector` that implements the asynchronous
  `select()`.
- Modify `DefaultClientRequestContext` and `ClientUtil` to use the new
  asynchronous `EndpointSelector.select()` if `selectNow()` returns
  `null`, so that it waits up to the connect timeout.
- Modify `UserClient` require `futureConverter` and `errorResponseFunction`
  at its construction time.
  - The method signatures of `protected execute()` have been changed.
- Make `AbstractListenable` more suitable for frequent updates, so we
  can add and remove many listeners when an `EndpointGroup` is empty.
- Miscellaneous:
  - Update `web-licenses.txt`

Result:

- Fixes #1910
- (Breaking) `EndpointSelector` API has been changed.
  - It is not a functional interface anymore.
  - `select()` has been renamed to `selectNow()`.
  - An asynchronous `select()` has been added.
- (Breaking) `UserClient` API has been changed.
  - Its constructor requires more parameters.
  - `execute()` requires less parameters.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java [hunk 1]
```diff
@@ -39,41 +42,77 @@ O initContextAndExecuteWithFallback(
...
+                return initContextAndExecuteWithFallback(delegate, ctx, errorResponseFactory, success);
+            } else {
+                return futureConverter.apply(initFuture.handle((success, cause) -> {
+                    try {
+                        if (cause != null) {
+                            throw cause;
+                        }
+
+                        return initContextAndExecuteWithFallback(
+                                delegate, ctx, errorResponseFactory, success);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java:75]: Could merge two lines?

File: core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java [hunk 2]
```diff
@@ -39,41 +42,77 @@ O initContextAndExecuteWithFallback(
...
-                if (res instanceof StreamMessage) {
-                    ((StreamMessage<?>) res).abort();
+            final CompletableFuture<Boolean> initFuture = ctx.init(endpointGroup);
+            if (initFuture.isDone()) {
+                // Initialization has been done immediately.
+                final boolean success;
+                try {
+                    success = initFuture.get();
+                } catch (Exception e) {
+                    throw Exceptions.peel(e);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java:63]: Wrap with `UnprocessedRequestException.wrap(e)` once #2836 is merged?

File: core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java [hunk 3]
```diff
@@ -39,41 +42,77 @@ O initContextAndExecuteWithFallback(
...
+
+            // We will use the fallback response which is created from the exception
+            // raised in ctx.init(), so the response returned can be aborted.
+            if (res instanceof StreamMessage) {
+                ((StreamMessage<?>) res).abort();
+            }
+
+            // No need to call `fail()` because failed by `DefaultRequestContext.init()` already.
+            final Throwable cause = ctx.log().partial().requestCause();
+            assert cause != null;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java:114]: Not related to this PR. Could we propagrate the `cause` to `res` via `res.abort(cause)`?

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java [hunk 4]
```diff
@@ -194,49 +199,84 @@ private static ServiceRequestContext serviceRequestContext() {
...
-            final ReleasableHolder<EventLoop> releasableEventLoop =
-                    options().factory().acquireEventLoop(sessionProtocol(), endpointGroup, endpoint);
-            eventLoop = releasableEventLoop.get();
-            log.whenComplete().thenAccept(unused -> releasableEventLoop.release());
+    private UnmodifiableFuture<Boolean> initEndpoint(Endpoint endpoint) {
+        this.endpointGroup = null;
+        updateEndpoint(endpoint);
+        runThreadLocalContextCustomizers();
+        acquireEventLoop(endpoint);
+        return UnmodifiableFuture.completedFuture(true);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:225]: How about making this private static final and reusing it?

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java [hunk 5]
```diff
@@ -194,49 +199,84 @@ private static ServiceRequestContext serviceRequestContext() {
...
+            acquireEventLoop(endpointGroup);
+            return UnmodifiableFuture.completedFuture(true);
         }
 
-        if (cause != null) {
-            failEarly(cause);
-            return false;
-        }
+        // Use an arbitrary event loop for asynchronous Endpoint selection.
+        final EventLoop temporaryEventLoop = options().factory().eventLoopSupplier().get();
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:242]: Can't we do `acquireEventLoop(endpointGroup);` first and use the eventLoop?

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java [hunk 6]
```diff
@@ -194,49 +199,84 @@ private static ServiceRequestContext serviceRequestContext() {
...
-        return true;
+            final EventLoop acquiredEventLoop = eventLoop();
+            if (acquiredEventLoop == temporaryEventLoop) {
+                // We were lucky. No need to hand over to other EventLoop.
+                return UnmodifiableFuture.completedFuture(success);
+            } else {
+                // We need to hand over to the acquired EventLoop.
+                return CompletableFuture.supplyAsync(() -> success, acquiredEventLoop);
+            }
+        }).thenCompose(Function.identity());
```
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:263]: To make `CompletableFuture<CompletableFuture<Boolean>>` to `CompletableFuture<Boolean>`?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        // Schedule the timeout task.
+        final ScheduledFuture<?> timeoutFuture =
+                executor.schedule(() -> listeningFuture.complete(null),
+                                  timeoutMillis, TimeUnit.MILLISECONDS);
+        listeningFuture.timeoutFuture = timeoutFuture;
+
+        // Cancel the timeout task if necessary, just in case listeningFuture is
+        // completed even before we assign timeoutFuture to listeningFuture.timeoutFuture.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java:84]: If so `listeningFuture.timeoutFuture = timeoutFuture;` this line should move down.
However, I'm wondering if there's a chance that `listeningFuture.isDone()` is `true` after we call `endpoint = selectNow(ctx);` when `endpoint` is `null`. 🤔

File: core/src/main/java/com/linecorp/armeria/internal/common/util/IdentityHashStrategy.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.internal.common.util;
+
+import it.unimi.dsi.fastutil.Hash.Strategy;
+
+public final class IdentityHashStrategy<T> implements Strategy<T> {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/common/util/IdentityHashStrategy.java:20]: Can we make this as an enum? 🤔

File: core/src/test/java/com/linecorp/armeria/client/ClientRequestContextDelayedInitTest.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2019 LINE Corporation
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/ClientRequestContextDelayedInitTest.java:2]: 2020?

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java (hunk 4 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:225]: It's reused inside `UnmodifiableFuture`.

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java (hunk 5 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:242]: We can, but why?

File: core/src/main/java/com/linecorp/armeria/internal/common/util/IdentityHashStrategy.java (hunk 8 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/common/util/IdentityHashStrategy.java:20]: Then we will lose the type parameter, i.e. the caller has to do an unchecked cast.

File: core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java (hunk 3 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java]: Done :-)

File: core/src/main/java/com/linecorp/armeria/internal/common/util/IdentityHashStrategy.java (hunk 8 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/common/util/IdentityHashStrategy.java:20]: Oops, I missed that. 😓

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java (hunk 4 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:225]: Ah I missed that as well. 😅

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java (hunk 7 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java]: Updated the comment to explain better.

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java (hunk 5 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:242]: Forget about this. 😅

File: core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java (hunk 2 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java]: Done.

**Comment by trustin** [comment]
@adriancole After this pull request is merged, a request to an empty endpoint group will wait up to connect timeout (3.2 seconds by default). If we are fine with failing the request if the endpoint group does not have any healthy endpoint after timeout, you can remove that code completely. Do you think we need to use a different timeout than the connect timeout, or do you have any other suggestions?

**Comment by adriancole** [comment]
connect timeout makes sense and is easiest to reason with

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:263]: Ah, flatten! 🤣

**Comment by trustin** [comment]
Thanks for reviewing!

**Pull request merged**